### ISSUE TITLE
RFC: image, video and audio generation from diffusion GGUFs (LTX-2)

### ISSUE THREAD

**Issue by ericcurtin** [ggml-org/llama.cpp#28541]
### Prerequisites

- [x] I am running the latest code. Mention the version if possible as well.
- [x] I carefully followed the [README.md](https://github.com/ggml-org/llama.cpp/blob/master/README.md).
- [x] I searched using keywords relevant to my issue to make sure that I am creating a new issue that is not already open (or closed).
- [x] I reviewed the [Discussions](https://github.com/ggml-org/llama.cpp/discussions), and have a new and useful enhancement to share.

### Feature Description

Native image, video and audio generation from latent diffusion transformers distributed as GGUF, starting with [LTX-2.3](https://huggingface.co/unsloth/LTX-2.3-GGUF) (22B audio-video DiT), so that

```sh
llama-server -hf unsloth/LTX-2.3-GGUF
```

works out of the box and serves OpenAI compatible `POST /v1/images/generations`, `POST /v1/videos` (+ `GET /v1/videos/{id}/content`) and `POST /v1/audio/speech`.

An implementation exists in #28540 (opened before this RFC, the PR checker asked for prior discussion). This issue is for agreeing on whether and how this should land.

### Motivation

- GGUF quantizations of diffusion transformers are becoming common (unsloth, city96, ...) and they run on ggml today only through stable-diffusion.cpp, which needs a ggml fork and its own server.
- LTX-2 conditions on *all* hidden states of a Gemma 3 text encoder, i.e. an LLM llama.cpp already runs well; the video/audio side is a DiT plus two VAEs, all expressible with existing ggml ops (conv 2d/3d via im2col, group/rms norm, flash attention, timestep embedding).
- One server binary for text, vision input and media output is what downstream tools (e.g. llmman CLI) want to launch.

### Possible Implementation

As in #28540:

- `tools/mediagen`: a library next to `tools/mtmd` (GGUF/safetensors loading, DiT graph, flow-matching Euler sampler, video VAE, audio VAE + vocoder, PNG/WAV/mp4 output) and a `llama-mediagen` CLI. The text encoder is a normal libllama model; its hidden states are captured through `cb_eval`, prompt enhancement reuses the same context in logits mode.
- `common`: `-hf` treats a repo with VAE / text projection sidecars as a diffusion model, downloads the sidecars and the text encoder repo of the model family; `--vae`, `--audio-vae`, `--text-proj`, `--text-encoder[-hf]`.
- `server`: when the model is a diffusion model, the media endpoints are served by the library and the text endpoints return not-supported.

Open questions for maintainers:

1. Is media generation in scope for llama.cpp at all, or should this stay a fork / separate project?
2. If in scope: a separate `tools/mediagen` library (as proposed) vs. extending libmtmd (which the mtmd dev guide funnels audio generation into).
3. Granularity for review: the PR could be split into (a) `common` sidecar resolution, (b) the library + CLI, (c) the server endpoints, plus two independent ggml fixes it currently works around (`ggml_conv_1d` output layout for batch > 1 with several output channels; CUDA `pad` mapping `ne1` to `gridDim.y`).
4. The `-hf` default text encoder per model family (`ltx-2 -> ggml-org/gemma-3-12b-it-GGUF`) is model-specific policy in `common`; alternatives are a `preset.ini` in the model repo or requiring `--text-encoder-hf`.

Sample outputs and numerical validation against the reference PyTorch implementation are in #28540.

**Comment by pwilkin** [comment]
Why would we want to add image generation to llama.cpp when stable-diffusion.cpp has that area covered?

LTX-2.3 is an old version of the model BTW, if we were adding support, it would make much more sense to go with LTX-2.5.

**Comment by ericcurtin** [comment]
> Why would we want to add image generation to llama.cpp when stable-diffusion.cpp has that area covered?

I think we want an OpenAI-compatible server that serves text, image, audio and video output. One could do a wrapper around llama.cpp and stable-diffusion.cpp but that feels like a bit of a workaround to creating a completely functional inference server. We also cut down on duplication of effort and de-duplicating code in general by making llama.cpp a more complete inference server. I would also propose inviting stable-diffusion.cpp into the ggml/llama.cpp community if they would like to.

> 
> LTX-2.3 is an old version of the model BTW, if we were adding support, it would make much more sense to go with LTX-2.5.

2 reasons I went with LTX-2.3: low hardware requirements and the other is unsloth only publish GGUFs for LTX-2.3, I don't trust all publishers, but I trust unsloth. Tagging @danielhanchen and @shimmyshimmer if they want to publish LTX-2.5 GGUFs and general conversation.

**Comment by pwilkin** [comment]
But LTX-2.3 and LTX-2.5 are exactly the same when it comes to requirements: 2.3 uses a 22B model + a 12B Gemma 3, while 2.5 uses a 22B model + a 12B Gemma 4.

**Comment by ericcurtin** [comment]
> But LTX-2.3 and LTX-2.5 are exactly the same when it comes to requirements: 2.3 uses a 22B model + a 12B Gemma 3, while 2.5 uses a 22B model + a 12B Gemma 4.

Fair enough, happy to work with LTX-2.5 when GGUFs are published :)

**Comment by ericcurtin** [comment]
The more I look around the more I think cosmos3-edge is actually the interesting one. But both ltx-2.5 and cosmos3-edge are interesting

**Comment by jacekpoplawski** [comment]
Do I understand correctly that the equivalent of a ComfyUI workflow graph is currently hardcoded in C++?

Do you plan to make the pipeline configurable?
